README.md

JK completed the excellent Google Python Exercises on October 5, 2015.

Google's Python Class: https://developers.google.com/edu/python/

Used the online IDE codeio here:
	https://codio.com/johnfkraus/google-python-exercises

Completed exercises are here:
	https://github.com/johnfkraus/google-python-exercises

Did not bother to clean up the code too much.  Got it to work and that was about it.

The Google Python tutorial was excellent IMHO.
 
Spoiler alert!

Image results from my logpuzzle solution:

![alt text](https://github.com/johnfkraus/google-python-exercises/blob/master/logpuzzle/animal_results_screenshot.png "animal log puzzle result screenshot")

![alt text][place]

[place]: https://github.com/johnfkraus/google-python-exercises/blob/master/logpuzzle/place_puzzle_solution_image.PNG "place log puzzle result screenshot"


elasticsearch/history.txt
   44  ll
   45  mv -nv ~/Downloads/mp4b/* .
   46  ls ~/Downloads/mp4b/
   47  cd ~/Documents
   48  ll
   49  cd receipts2/
   50  l
   51  ll
   52  cd mp4/
   53  ll
   54  pwd
   55  pwd
   56  cd /Volumes/LACIE\ SHARE/Documents/videomp4/
   57  ll
   58  mv -nv ~/Documents/receipts2/mp4/* .
   59  ll ~/Documents/receipts2/mp4
   60  cd ../
   61  ll
   62  mkdir duplicate-videos-maybe
   63  cd duplicate-videos-maybe/
   64  mv ~/Documents/receipts2/mp4/* .
   65  cd ~/Documents/receipts2/
   66  ll
   67  tree .
   68  rmdir mp4/
   69  cd mp4/
   70  ll
   71  cd ../
   72  rm -rf mp4
   73  ll
   74  cd ../
   75  ll
   76  rm -rf receipts2/
   77  ll
   78  df -h
   79  exit
   80  cd Downloads/
   81  ll
   82  rm *.crdownload
   83  ll
   84  ll
   85  ll
   86  df -h
   87  df -h
   88  df -h
   89  df -h
   90  df -h
   91  df -h
   92  df -h
   93  df -h
   94  df -h
   95  df -h
   96  df -h
   97  df -h
   98  df -h
   99  df -h
  100  df -h
  101  df -h
  102  df -h
  103  df -h
  104  df -h
  105  df -h
  106  df -h
  107  df -h
  108  df -h
  109  df -h
  110  df -h
  111  df -h
  112  df -h
  113  exit
  114  sudo su
  115  man srm
  116  srm --help
  117  which srm
  118  which gshred
  119  man gshred
  120  df -h
  121  df -h
  122  df -h
  123  df -h
  124  df -h
  125  df -h
  126  df -h
  127  df -h
  128  df -h
  129  df -h
  130  cd Downloads/
  131  l
  132  ll
  133  mv -nv *.mp4 /Volumes/LACIE\ SHARE/Documents/videomp4/
  134  mv -nv *.mp4 /Volumes/LACIE\ SHARE/Documents/videomp4/
  135  ll
  136  rm Cute\ Skinny\ Teen\ Playing\ have\ Fun\ with\ Ohmibod\ in\ Pussy\ on\ Webcam\ -\ Pornhubcom.mp4 
  137  rm Micro\ Bikini\ Babe\ Shows\ off\ in\ Public\ -\ Pornhubcom.mp4 
  138  rm Bikini-\ Pleasure\ -\ Pornhubcom.mp4 
  139  ll
  140  ls *.mp4
  141  rm Extreme\ Bikini\ Babe\ Kagney\ Linn\ Karter\ Pounds\ her\ Pussy\ -\ Pornhubcom.mp4 
  142  rm .com.google.Chrome.AN8X25 
  143  ll
  144  ll
  145  ll
  146  exit
  147  df -h
  148  df -h
  149  cd Downloads/
  150  ll
  151  mv *.mp4 mp4b/
  152  ll
  153  tree mp4b/
  154  cd mp4b/
  155  ll
  156  rm Clip\ \(August\ *
  157  ll
  158  ll
  159  mkdir messervideos
  160  ll
  161  rmdir messervideos/
  162  cd ../
  163  ll
  164  mkdir messervideos
  165  mv sy0-501-section0-video.zip messervideos/
  166  ll
  167  ll
  168  mv sy0-501-section3-video.zip messervideos/
  169  ll
  170  cd messervideos/
  171  ll
  172  unzip sy0-501-section0
  173  unzip sy0-501-section0-video.zip 
  174  ll
  175  unzip sy0-501-section3-video.zip 
  176  ll
  177  ll
  178  unzip sy0-501-section1-video.zip 
  179  unzip sy0-501-section2-video.zip 
  180  ll
  181  unzip sy0-501-section4-video.zip 
  182  unzip sy0-501-section5-video.zip 
  183  unzip sy0-501-section6-video.zip 
  184  ls -lt
  185  ls -l
  186  ls *.srt
  187  cat sy0-501-0001-01-how-to-pass-your-security-plus-exam-v1.srt 
  188  ll
  189  mkdir zipmesser
  190  mv *.zip zipmesser/
  191  ll
  192  mv *.srt zipmesser/
  193  ll
  194  df -h
  195  cd ../
  196  ll
  197  mkdir toBeFiledOrArchived
  198  cd /Volumes/
  199  ll
  200  cd UBUNTU_103/
  201  ll
  202  cd TRIPLEK/
  203  ll
  204  cd Logs/
  205  ll
  206  cd ../
  207  ll
  208  cd cactus/
  209  ll
  210  cd ../
  211  ll
  212  cp -rf Logs ~/workspaces/
  213  cd john
  214  ll
  215  tree . | more
  216  ll
  217  cd anna/
  218  ll
  219  cd turbotax-anna/
  220  ll
  221  tree . | more
  222  ll
  223  cd turbotax
  224  ll
  225  file .
  226  ll
  227  file
  228  file *
  229  cd ../
  230  ll
  231  cd ../
  232  ll
  233  cd ../
  234  ll
  235  cd nuwave/
  236  ll
  237  cd /Volumes/UBUNTU_103/
  238  ll
  239  df -h
  240  exit
  241  docker pull docker.elastic.co/elasticsearch/elasticsearch:6.3.2
  242  exit
  243  vim installettercap.txt
  244  ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)" < /dev/null 2> /dev/null
  245  df -h
  246  brew install ettercap
  247  ettercap
  248  man ettercap
  249  man ettercap
  250  ettercap -i
  251  ettercap --help
  252  ettercap -G
  253  ettercap -P list
  254  ll
  255  docker pull docker.elastic.co/kibana/kibana:6.3.2
  256  cd /etc
  257  ll
  258  cd /etc
  259  ll
  260  which docker-compose
  261  exit
  262  screen ~/Library/Containers/com.docker.docker/Data/com.docker.driver.amd64-linux/tty
  263  ettercap -I
  264  git status
  265  git init
  266  git status
  267  git add --all
  268  git commit -m 'init'
  269  git checkout -b develop
  270  git status
  271  ll
  272  cd johnkraus/
  273  ll
  274  cd src/
  275  ll
  276  cd kotlininaction/
  277  ll
  278  kotlinc Hello.kt 
  279  which koton
  280  which kotlin
  281  ll
  282  cd src/geometry/
  283  ll
  284  cd shapes/
  285  ll
  286  rmdir example/
  287  ll
  288  cd ../
  289  ll
  290  ll
  291  cd shapes/
  292  ll
  293  mv Rectangle.kt ../
  294  ll
  295  cd ../
  296  ll
  297  cd example/
  298  ll
  299  mv Example.kt ../
  300  ll
  301  cd ../
  302  ll
  303  ll
  304  rmdir example/
  305  ll
  306  ll
  307  rm shapes/
  308  rmdir shapes/
  309  ll
  310  git status
  311  cd ../
  312  ll
  313  cd ../
  314  ll
  315  git status
  316  git init
  317  git add --all
  318  git status
  319  git status
  320  git status
  321  git add --all
  322  git commit -m 'init'
  323  git status
  324  git checkout -b develop
  325  git status
  326  vim .gitignore 
  327  git status
  328  git add --all
  329  git commit -m 'ch 2'
  330  git status
  331  ncat --exec "/bin/bash" -l 8081 --keep-open
  332  ettercap
  333  arp -a
  334  netstat
  335  df -h
  336  df -h
  337  df -h
  338  df -h
  339  df -h
  340  df -h
  341  df -h
  342  df -h
  343  df -h
  344  ll
  345  ws
  346  ll
  347  cd kotlin-in-action/
  348  ll
  349  ll
  350  cd src/
  351  ll
  352  cp -rf ch* ~/workspaces/sandbox/src/
  353  ll
  354  ws
  355  ll
  356  cd sandbox/
  357  ll
  358  cd src/
  359  ll
  360  mkdir jk

basic/mimic.py
#!/usr/bin/python -tt
# Copyright 2010 Google Inc.
# Licensed under the Apache License, Version 2.0
# http://www.apache.org/licenses/LICENSE-2.0

# Google's Python Class
# http://code.google.com/edu/languages/google-python-class/

"""Mimic pyquick exercise -- optional extra exercise.
Google's Python Class

Read in the file specified on the command line.
Do a simple split() on whitespace to obtain all the words in the file.
Rather than read the file line by line, it's easier to read
it into one giant string and split it once.

Build a "mimic" dict that maps each word that appears in the file
to a list of all the words that immediately follow that word in the file.
The list of words can be be in any order and should include
duplicates. So for example the key "and" might have the list
["then", "best", "then", "after", ...] listing
all the words which came after "and" in the text.
We'll say that the empty string is what comes before
the first word in the file.

With the mimic dict, it's fairly easy to emit random
text that mimics the original. Print a word, then look
up what words might come next and pick one at random as
the next work.
Use the empty string as the first word to prime things.
If we ever get stuck with a word that is not in the dict,
go back to the empty string to keep things moving.

Note: the standard python module 'random' includes a
random.choice(list) method which picks a random element
from a non-empty list.

For fun, feed your program to itself as input.
Could work on getting it to put in linebreaks around 70
columns, so the output looks better.

"""

import random
import sys

def print_dict(dict):
  for k, v in dict.items(): 
    print k, '>', v

def print_dict_count(dict):
  for k, v in dict.items(): 
    print k, '>', v, '; len(v) = ', len(v)

def mimic_dict(filename):
  """Returns mimic dict mapping each word to list of words which follow it."""
  # +++your code here+++
  db = False # debugging messages on?
  dict = {}
  file = open(filename, 'rU')
  contents = file.read()
  contents_list = contents.split()
  # print 'len = ', len(contents_list)
  dict[''] = [contents_list[0]]
  i = 0
  while i < len(contents_list)-1:
    if db: print 'line 59', print_dict(dict)
    word1 = contents_list[i]
    word2 = contents_list[i+1]
    if db: print 'line 60; i = ', i, '/', len(contents_list),'; ', word1,'->', word2
    if word1 in dict:
      if db: print 'line 62; i = ', i, '/', len(contents_list),'; ', word1,'->', word2, '; dict[', word1, '] = ', dict[word1]
      value_list = dict[word1]
      if db: print 'line 64; i = ', i, '/', len(contents_list),';', word1,'->', word2, '; value_list = ', value_list
      if db: print 'type(value_list) = ', type(value_list)
      # duplicates are not added to the values list, although this was not an
      # explicit direction; duplicates might appropriately reflect word use
      # probability; but watch out for repeating words in output!
      if word2 not in value_list: 
        value_list.append(word2)
        dict[word1] = value_list
    else:
      # word1 is not already in dict; add it
      dict[word1] = [word2]
      if db: print 'line 73; i = ', i, '/', len(contents_list),'; ', word1,'->', word2, '; dict[', word1, '] = ', dict[word1], 'type(dict[word1]) = ', type(dict[word1])
    i+=1
  if db: print dict
  if db: print 'len(dict) = ', len(dict)
  if db: print_dict(dict)
  if db: print_dict_count(dict)
  if db: print 'dict = ', dict
  return dict

def pick_random_from_list(list):
  # print '95 list = ', list
  if len(list) > 1:
    # print random.randrange(0, len(list)-1)
    return list[random.randrange(0, len(list)-1)]
  else:
  # return list(random.randrange(0, len(list)-1))
    return list[0]
  return

def print_mimic(mimic_dict, word):
  """Given mimic dict and start word, prints 200 random words."""
  # +++your code here+++
  # print word
  # list = mimic_dict[word]
  # print mimic_dict[word] # returns list
  i = 0
  string = ''
  while i < 200:
    list = mimic_dict[word]
    word = pick_random_from_list(list)
    # print word
    string = string + ' ' +  word
    i+=1
  print string
  return


# Provided main(), calls mimic_dict() and mimic()
def main():
  if len(sys.argv) != 2:
    print 'usage: ./mimic.py file-to-read'
    sys.exit(1)
  dict = mimic_dict(sys.argv[1])
  print_mimic(dict, '')


if __name__ == '__main__':
  main()
